ddclient.in

```diff
@@ -2146,6 +2154,128 @@ sub geturl {
+        fatal("You requested network access with curl but curl is not installed.");
+    }
+
+    ## canonify proxy and url
+    $force_ssl = 1 if ($url =~ /^https:/);
+    $proxy  =~ s%^https?://%%i;
+    $url    =~ s%^https?://%%i;
+    $server = $url;
+    $server =~ s%/.*%%;
+    $url    = "/" unless $url =~ m%/%;
+    $url    =~ s%^[^/]*/%%;
+
+    debug("opt(fw = %s)", opt('fw')) if opt('fw');
+    debug("glo fw = %s", $globals{'fw'}) if $globals{'fw'};
+    ## always omit SSL for connections to local router
+    $use_ssl = 1 if ($force_ssl || ($globals{'ssl'} && !$is_get_ip));
+
+    $protocol = ($use_ssl ? "https" : "http");
+
+    debug("proxy    = %s", $proxy);
+    debug("protocol = %s", $protocol);
+    debug("server   = %s", $server);
+    debug("url      = %s", $url);
+    debug("ip ver   = %s", $ipversion);
+
+    if (!opt('exec')) {
+        debug("skipped network connection");
+        verbose("SENDING:", "%s", "${server}/${url}");
+    } else {
+        push(@curlopt, "silent");
+        push(@curlopt, "insecure") if ($use_ssl && $is_get_ip);  ## this should really be a user option
+        push(@curlopt, "cacert=".opt('ssl_ca_file')) if defined(opt('ssl_ca_file'));
+        push(@curlopt, "capath=".opt('ssl_ca_dir')) if defined(opt('ssl_ca_dir'));
+        push(@curlopt, "ipv4") if ($ipversion eq '4');
+        push(@curlopt, "ipv6") if ($ipversion eq '6');
+        push(@curlopt, "user-agent=\"${program}/${version}\"");
+        push(@curlopt, "connect-timeout=$timeout");
+        push(@curlopt, "max-time=$timeout");
+        push(@curlopt, "request=$method");
+        push(@curlopt, "user=${login}:${password}") if (($login ne '') && ($password ne ''));
+        push(@curlopt, "proxy=\"${protocol}://${proxy}\"") if ($proxy ne '');
+        push(@curlopt, "url=\"${protocol}://${server}/${url}\"");
+        @header_lines = split('\n', $headers);
+        foreach (@header_lines) { s/"/\\"/g; $_ = "header=\"$_\"" };
+        push(@curlopt, @header_lines);
+        $data =~ s/"/\\"/g;  ## Escape double-quotes
+        push(@curlopt, "data=\"${data}\"") if ($data ne '');
+
+        # don't include ${url} as that might expose login credentials
+        $0 = sprintf("%s - curl sending to %s", $program, "${protocol}://${server}");
+        verbose("SENDING:", "curl to %s", "${protocol}://${server}");
+
+        $reply = curl_cmd(@curlopt);
+
+        if (!$reply) {
+            # don't include ${url} as that might expose login credentials
+            warning("curl cannot connect to %s://%s using IPv%s",${protocol},${server},$ipversion);
+        }
+    }
+
+    ## during testing simulate reading the URL
+    if (opt('test')) {
+        my $filename = "$server/$url";
+        $filename =~ s|/|%2F|g;
+        if (opt('exec')) {
+            $reply = save_file("$savedir/$filename", $reply, 'unique');
+        } else {
+            $reply = load_file("$savedir/$filename");
+        }
+    }
+
+    $reply =~ s/\r//g if defined $reply;
```

Comment:
Move common code to `geturl`.

Reply:
Agree.  But remember you were seeking simplicity of code review so I separated out the functions.  I actually think the curl and non-curl should get placed into a single geturl because there are common pieces that wrap both, but I don't want to have to create a new function parameter definitions (all the canonify stuff) that impact both curl and non-curl.

Reply:
I plan on using [URI](https://metacpan.org/pod/URI) to handle URL parsing, and that should minimize argument packing and unpacking between `geturl` and the `fetch_via_*` functions. Stay tuned.

Reply:
I'll wait and see what you do... Maybe you would like to implement the `geturl()` wrapper and `fetch_via_socket_io()` functions, then I can sync this pull request to that and insert the `fetch_via_curl()` part?